scipy/optimize/_differentialevolution.py

```diff
@@ -598,6 +627,41 @@ def __init__(self, func, bounds, args=(),
 
         self.random_number_generator = check_random_state(seed)
 
+        # Which parameters are going to be integers?
+        if np.any(integrality):
+            # user has provided a truth value for integer constraints
+            if polish:
+                message = ("Polishing is incompatible with integrality"
+                           " constraints; `polish` will be ignored.")
+                self.polish = False
+                warnings.warn(message, OptimizeWarning)
+
+            self.integrality = np.broadcast_to(
+                integrality,
+                self.parameter_count
+            )
+            self.integrality = np.asarray(self.integrality, bool)
+            # See if rounding would allow a parameter to strictly exceed
+            # the stated limits. For example using bounds of (0.2, 9.8) for an
+            # integral parameter will utilise integers in the range:
+            #       [np.round(0.2), np.round(9.8)] == [0, 10]
+            # The values 0 and 10 are strictly outside the applied bounds.
+            # A design choice was made to raise a warning, but to still
+            # proceed. This is because the situation may be relatively common.
+            # Alternatives could be to raise a RuntimeError.
```

Comment:
I think it would be better in this example to automatically adjust the bounds to (0.5, 9.5).

Reply:
What would be the rule for expanding or contracting the bounds?

- [0, 10] --> expand to [-0.5, 10.5]?
- [0.2, 9.8] --> expand to [-0.5, 10.5] or contract to [0.5, 9.5]? 
- [0.2, 0.9] ??? There are no included integers.

Would the rule to be to only include integers within the range, including the end points? This would involve raising an Error for the last example above.

- [0, 10] would change the limits to [-0.5, 10.5]
- [0.2, 9.8] would change the limits to [0.5, 9.5]
- [0.2, 0.9] would raise an Error.
- [-4.8, 2] would change the limits to [-4.5, 2.5]

Reply:
i.e. the limits become `[(np.fix(lb) - 0.5, np.fix(ub)+0.5)]`, with the exception for no possible value

Reply:
> [0, 10] --> expand to [-0.5, 10.5]?

Yes. Each integer 0-10 will have equal chance of being chosen when rounding a continuous value in that range, and you will not violate the user's specified bounds of 0-10.

> [0.2, 9.8] --> expand to [-0.5, 10.5] or contract to [0.5, 9.5]?

Ceil(0.2) - 0.5 = 0.5 and floor(9.8)+0.5 = 9.5, so contract to  [0.5, 9.5]. Each integer 1 - 9 will have equal chance of being chosen when rounding a continuous value in that range, and you will not violate the user's specified bounds of 1-9.

> [0.2, 0.9] ??? There are no included integers.

Gets modified to [0.5, 0.5]. 0.5 - 0.5 is not >= 1, so "raise an error that there are no solutions within bounds."

> [-4.8, 2] would change the limits to [-4.5, 2.5]

ceil(-4.8) -0.5 = -4.5 and floor(2) + 0.5 = 2.5, so yes, the limits change to [-4.5, 2.5].